#### Lisa Over
CMPINF 2110<br>
Homework 1<br>
January 26, 2021

In [2]:
import sys
print(sys.executable)

/opt/anaconda3/envs/cmpinf_2110/bin/python


In [3]:
import pandas as pd

## Women of 2020
This dataset contains information about 100 inspiring and influential women from all over the world who won an award in 2020. A data dictionary exists and provides the name of the six variables in the dataset, the data type of each variable (which are all character), and a description of each variable. I would feel comfortable accessing and using this dataset. Each record is one woman, and one of the variables is a description of that woman. I think it would be interesting to create a word map or to do a topic analysis with graphs to visualize the different achievements made by these women as described in the description column.

#### women.csv 
Read the raw data in from tidytuesday site.

In [4]:
data_url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-12-08/women.csv'

In [5]:
women_df = pd.read_csv(data_url)

#### Data description
There are 100 rows, one for each woman, and six variables: name, img (link to image), category, country, role, and description. The variables are consistent with the Data Dictionary. All variables are pandas strings: Dtype object or dtype('O'). There are no missing values in this dataset. 

In [6]:
women_df.shape

(100, 6)

In [7]:
women_df.head()

name                                                img  \
0         Unsung hero  https://ichef.bbci.co.uk/news/300/cpsprodpb/F2...   
1  Loza Abera Geinore  https://ichef.bbci.co.uk/news/300/cpsprodpb/15...   
2         Houda Abouz  https://ichef.bbci.co.uk/news/300/cpsprodpb/14...   
3     Christina Adane  https://ichef.bbci.co.uk/news/300/cpsprodpb/4A...   
4  Yvonne Aki-Sawyerr  https://ichef.bbci.co.uk/news/300/cpsprodpb/13...   

     category       country                 role  \
0         All     Worldwide  Making a difference   
1  Leadership      Ethiopia           Footballer   
2  Creativity       Morocco               Rapper   
3  Leadership   Netherlands           Campaigner   
4  Leadership  Sierra Leone                Mayor   

                                         description  
0  In an extraordinary year, when countless women...  
1  Loza Abera Geinore was born and raised in a sm...  
2  Houda Abouz, aka Khtek, is a Moroccan rapper k...  
3  Christina was behind the UK petition for free ...  
4  Mayor Yvonne Aki-Sawyerr OBE is best known for...

Display missing observations and data types for each variable.

In [8]:
women_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         100 non-null    object
 1   img          100 non-null    object
 2   category     100 non-null    object
 3   country      100 non-null    object
 4   role         100 non-null    object
 5   description  100 non-null    object
dtypes: object(6)
memory usage: 4.8+ KB


In [9]:
print(type(women_df.name[0]))

<class 'str'>


In [10]:
women_df.name.dtype

dtype('O')

Display number of unique observations per variable and display value counts for certain variables.

In [11]:
women_df.nunique()

name           100
img            100
category         5
country         67
role            75
description    100
dtype: int64

In [12]:
women_df.category.value_counts()

Knowledge     32
Leadership    29
Creativity    21
Identity      17
All            1
Name: category, dtype: int64

In [13]:
women_df.country.value_counts()

US          9
UK          6
Thailand    3
India       3
Nigeria     2
           ..
Nepal       1
India       1
Ecuador     1
Mexico      1
Belarus     1
Name: country, Length: 67, dtype: int64

In [14]:
women_df.role.value_counts()

Activist                6
Disability activist     4
Campaigner              4
Scientist               3
Musician                3
                       ..
Union leader            1
Rapper                  1
Global health leader    1
Archaeologist           1
Waorani leader          1
Name: role, Length: 75, dtype: int64

## European energy
This entry contains two files that provide information about energy use in Europe: energy_types and country_totals. There is a data dictionary for each dataset, but there is no difference between the variables and variable descriptions. They each specify the same seven variables with the same datatypes and descriptions. The only difference is that each file has a different statement before the data dictionary table. I feel comfortable accessing this data and actually am interested in finding out what is in each file and how the files relate to each other, if at all. 

#### energy_types.csv 
Read the raw data in from tidytuesday site.

In [15]:
data_url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-08-04/energy_types.csv'

In [16]:
energy_df = pd.read_csv(data_url)

#### Data description
There are 296 rows and seven variables: country (2-character abbreviation), country_name, type, level, 2016, 2017, 2018. Country, country_name, and type are pandas strings, and the year columns are float.

The statement in the data dictionary for this file is, "Limited to Level 1 or 2 production by type - either conventional thermal (fossil fuels), nuclear, hydro, wind, solar, geothermal, or other." After looking at the file, this statement is referring to the type column, which indicates the energy type as one of eight types: solar, conventional thermal, wind, pumped hydro power, other, hydro, geothermal, and nuclear. This data has country totals for the eight energy types for three years from 2016 to 2018. I would rename this file to be country_totals_by_energy_type. I would also move the information from the note into the table. The description of the level variable already specifies level 1 or 2, but the description of the type variable could specify that it is a type of energy and list the possible energy type values.

There are only eight missing values in this dataset, all in the country_name column, which could probably be filled in easily because the country field is not missing.  

In [17]:
energy_df.shape

(296, 7)

In [18]:
energy_df.head()

country country_name                  type    level     2016     2017  \
0      BE      Belgium  Conventional thermal  Level 1  30728.0  31316.0   
1      BE      Belgium               Nuclear  Level 1  41430.0  40128.5   
2      BE      Belgium                 Hydro  Level 1   1476.0   1360.9   
3      BE      Belgium    Pumped hydro power  Level 2   1110.0   1093.2   
4      BE      Belgium                  Wind  Level 1   5340.0   6387.9   

        2018  
0  30092.635  
1  26995.628  
2   1239.248  
3    983.190  
4   7177.346

Display missing observations and data types for each variable.

In [19]:
energy_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 296 entries, 0 to 295
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   country       296 non-null    object 
 1   country_name  288 non-null    object 
 2   type          296 non-null    object 
 3   level         296 non-null    object 
 4   2016          296 non-null    float64
 5   2017          296 non-null    float64
 6   2018          296 non-null    float64
dtypes: float64(3), object(4)
memory usage: 16.3+ KB


In [20]:
print(energy_df.isna().sum())

country         0
country_name    8
type            0
level           0
2016            0
2017            0
2018            0
dtype: int64


In [34]:
missing_df = energy_df.loc[energy_df['country_name'].isna()]
missing_df

country country_name                  type    level        2016  \
216      UK          NaN  Conventional thermal  Level 1  203165.188   
217      UK          NaN               Nuclear  Level 1   65149.080   
218      UK          NaN                 Hydro  Level 1    8286.510   
219      UK          NaN    Pumped hydro power  Level 2    2948.890   
220      UK          NaN                  Wind  Level 1   37262.699   
221      UK          NaN                 Solar  Level 1   10410.948   
222      UK          NaN            Geothermal  Level 1       0.000   
223      UK          NaN                 Other  Level 1       0.000   

           2017        2018  
216  189296.528  180837.905  
217   63886.860   59097.753  
218    8723.430    7678.659  
219    2862.010    2515.998  
220   50003.654   56903.961  
221   11524.870   12857.349  
222       0.000       0.000  
223       0.000       0.000

Display number of unique observations per variable and display value counts for certain variables.

In [35]:
energy_df.nunique()

country          37
country_name     36
type              8
level             2
2016            194
2017            203
2018            184
dtype: int64

In [36]:
energy_df.type.value_counts()

Nuclear                 37
Conventional thermal    37
Solar                   37
Geothermal              37
Hydro                   37
Wind                    37
Pumped hydro power      37
Other                   37
Name: type, dtype: int64

#### country_totals .csv
Read the raw data in from tidytuesday site.

In [37]:
data_url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-08-04/country_totals.csv'

In [38]:
country_df = pd.read_csv(data_url)

#### Data description
There are 185 rows and seven variables: country (2-character abbreviation), country_name, type, level, 2016, 2017, 2018. Country, country_name, and type are pandas strings, and the year columns are float.

The statement in the data dictionary for this file is, "Limited to total net production, along with imports, exports, energy lost, and energy supplied (net + import - export - energy absorbed by pumping)." This statement is referring to the type column, which indicates the energy production type with the following values: energy absorbed by pumping, total net production, exports, energy supplied, and imports. This data has country totals for the eight energy production types for three years from 2016 to 2018. I would have named this file country_totals_by_production_type. I would also move the information from the note into the table. The description of the level variable already specifies level 1 or 2, but the description of the type variable could specify that it is energy production type and list the possible energy production values.

There are six missing values in this dataset, five in the country_name column and one in the 2016 column. It is probably possible to fill in the missing country_name values because the country field has no missing values. And, because energy gain (total net production + imports) minus energy loss (exports + energy absorbed by pumping) equals the total net production gain, I think the missing 2016 value could be calculated from the other values for that country.  

In [39]:
country_df.shape

(185, 7)

In [40]:
country_df.head()

country country_name                        type  level     2016     2017  \
0      BE      Belgium        Total net production  Total  82520.0  82948.5   
1      BE      Belgium                     Imports  Total  14648.0  14189.4   
2      BE      Belgium                     Exports  Total   8465.0   8167.8   
3      BE      Belgium  Energy absorbed by pumping  Total   1475.0   1485.4   
4      BE      Belgium             Energy supplied  Total  87228.0  87484.7   

        2018  
0  69212.347  
1  21635.908  
2   4308.347  
3   1347.901  
4  85192.007

Display missing observations and data types for each variable.

In [41]:
country_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 185 entries, 0 to 184
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   country       185 non-null    object 
 1   country_name  180 non-null    object 
 2   type          185 non-null    object 
 3   level         185 non-null    object 
 4   2016          184 non-null    float64
 5   2017          185 non-null    float64
 6   2018          185 non-null    float64
dtypes: float64(3), object(4)
memory usage: 10.2+ KB


In [42]:
print(country_df.isna().sum())

country         0
country_name    5
type            0
level           0
2016            1
2017            0
2018            0
dtype: int64


In [47]:
missing_df = country_df.loc[country_df['country_name'].isna()]
missing_df

country country_name                        type  level        2016  \
135      UK          NaN        Total net production  Total  324274.434   
136      UK          NaN                     Imports  Total   20017.923   
137      UK          NaN                     Exports  Total    2272.866   
138      UK          NaN  Energy absorbed by pumping  Total    4014.340   
139      UK          NaN             Energy supplied  Total  338005.151   

           2017        2018  
135  323439.536  317375.629  
136   18166.980   21332.415  
137    3407.049    2224.764  
138    3859.190    3390.866  
139  334340.277  333092.414

Display number of unique observations per variable and display value counts for certain variables.

In [48]:
country_df.nunique()

country          37
country_name     36
type              5
level             1
2016            167
2017            169
2018            169
dtype: int64

In [49]:
country_df.type.value_counts()

Energy absorbed by pumping    37
Energy supplied               37
Total net production          37
Imports                       37
Exports                       37
Name: type, dtype: int64

## Traumatic Brain Injury
This entry contains four files about traumatic brain injuries sustained by civilians and military personnel. Three of the files, tbi_age.csv, tbi_military.csv, and tbi_year.csv, each have a data dictionary that specifies the variable name, data type, and description. Each of these files has five variables, three character and two double. The file dod_tbi_2006.csv does not have a data dictionary. I feel compfortable accessing this data.

#### tbi_age.csv 
Read the raw data in from tidytuesday site.

In [50]:
data_url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-03-24/tbi_age.csv'

In [51]:
age_df = pd.read_csv(data_url)

#### Data description
This file specifies the number of TBIs and the rate per 100,000 by age group, type of measure, and injury mechanism. There are 231 observations and five variables: age_group, type (type of measure), injury_mechanism, number_est, and rate_est. The variable descriptions for number_est and rate_est in the data dictionary indicate that this data is from 2014.

There are 11 missing values for number_est and 11 missing values for rate_est. These missing values occur in the same rows, i.e., number_est is missing in the same rows that rate_est is missing, and only in rows where the injury mechanism is 'Intentional self-harm.'

In [52]:
age_df.shape

(231, 5)

In [53]:
age_df.head(20)

age_group                        type  \
0       0-17  Emergency Department Visit   
1       0-17  Emergency Department Visit   
2       0-17  Emergency Department Visit   
3       0-17  Emergency Department Visit   
4       0-17  Emergency Department Visit   
5       0-17  Emergency Department Visit   
6       0-17  Emergency Department Visit   
7        0-4  Emergency Department Visit   
8        0-4  Emergency Department Visit   
9        0-4  Emergency Department Visit   
10       0-4  Emergency Department Visit   
11       0-4  Emergency Department Visit   
12       0-4  Emergency Department Visit   
13       0-4  Emergency Department Visit   
14      5-14  Emergency Department Visit   
15      5-14  Emergency Department Visit   
16      5-14  Emergency Department Visit   
17      5-14  Emergency Department Visit   
18      5-14  Emergency Department Visit   
19      5-14  Emergency Department Visit   

                                     injury_mechanism  number_est  rate_est  
0                               Motor Vehicle Crashes     47138.0      64.1  
1                                 Unintentional Falls    397190.0     539.8  
2      Unintentionally struck by or against an object    229236.0     311.6  
3   Other unintentional injury, mechanism unspecified     55785.0      75.8  
4                               Intentional self-harm         NaN       NaN  
5                                             Assault     24360.0      33.1  
6                     Other or no mechanism specified     57983.0      78.8  
7                               Motor Vehicle Crashes      5464.0      27.5  
8                                 Unintentional Falls    230776.0    1161.0  
9      Unintentionally struck by or against an object     53436.0     268.8  
10  Other unintentional injury, mechanism unspecified     12007.0      60.4  
11                              Intentional self-harm         NaN       NaN  
12                                            Assault       674.0       3.4  
13                    Other or no mechanism specified     19360.0      97.4  
14                              Motor Vehicle Crashes     19785.0      48.0  
15                                Unintentional Falls    133084.0     323.1  
16     Unintentionally struck by or against an object    120839.0     293.4  
17  Other unintentional injury, mechanism unspecified     30656.0      74.4  
18                              Intentional self-harm         NaN       NaN  
19                                            Assault      9690.0      23.5

Display missing observations and data types for each variable.

In [54]:
age_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231 entries, 0 to 230
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   age_group         231 non-null    object 
 1   type              231 non-null    object 
 2   injury_mechanism  231 non-null    object 
 3   number_est        220 non-null    float64
 4   rate_est          220 non-null    float64
dtypes: float64(2), object(3)
memory usage: 9.1+ KB


In [55]:
print(age_df.isna().sum())

age_group            0
type                 0
injury_mechanism     0
number_est          11
rate_est            11
dtype: int64


Display number of unique observations per variable and display value counts for certain variables.

In [56]:
age_df.nunique()

age_group            11
type                  3
injury_mechanism      7
number_est          216
rate_est            148
dtype: int64

In [57]:
age_df.age_group.value_counts()

25-34    21
5-14     21
55-64    21
65-74    21
15-24    21
0-17     21
Total    21
75+      21
0-4      21
45-54    21
35-44    21
Name: age_group, dtype: int64

Look at rows for one age group to verify age data is provided by age, type, and injury mechanism.

In [58]:
age0_17_df = age_df.loc[age_df['age_group'] == '0-17']

In [59]:
age0_17_df.shape

(21, 5)

In [60]:
age0_17_df.head(21)

age_group                        type  \
0        0-17  Emergency Department Visit   
1        0-17  Emergency Department Visit   
2        0-17  Emergency Department Visit   
3        0-17  Emergency Department Visit   
4        0-17  Emergency Department Visit   
5        0-17  Emergency Department Visit   
6        0-17  Emergency Department Visit   
77       0-17            Hospitalizations   
78       0-17            Hospitalizations   
79       0-17            Hospitalizations   
80       0-17            Hospitalizations   
81       0-17            Hospitalizations   
82       0-17            Hospitalizations   
83       0-17            Hospitalizations   
154      0-17                      Deaths   
155      0-17                      Deaths   
156      0-17                      Deaths   
157      0-17                      Deaths   
158      0-17                      Deaths   
159      0-17                      Deaths   
160      0-17                      Deaths   

                                      injury_mechanism  number_est  rate_est  
0                                Motor Vehicle Crashes     47138.0      64.1  
1                                  Unintentional Falls    397190.0     539.8  
2       Unintentionally struck by or against an object    229236.0     311.6  
3    Other unintentional injury, mechanism unspecified     55785.0      75.8  
4                                Intentional self-harm         NaN       NaN  
5                                              Assault     24360.0      33.1  
6                      Other or no mechanism specified     57983.0      78.8  
77                               Motor Vehicle Crashes      5830.0       7.9  
78                                 Unintentional Falls      7935.0      10.8  
79      Unintentionally struck by or against an object      1985.0       2.7  
80   Other unintentional injury, mechanism unspecified      3135.0       4.3  
81                               Intentional self-harm         NaN       NaN  
82                                             Assault      1535.0       2.1  
83                     Other or no mechanism specified      2595.0       3.5  
154                              Motor Vehicle Crashes       931.0       1.3  
155                                Unintentional Falls        48.0       0.1  
156     Unintentionally struck by or against an object        52.0       0.1  
157  Other unintentional injury, mechanism unspecified       308.0       0.4  
158                              Intentional self-harm         NaN       NaN  
159                                            Assault       611.0       0.8  
160                    Other or no mechanism specified        61.0       0.1

Look at missing values to determine if there is a pattern.

In [62]:
missing_df = age_df.loc[age_df['number_est'].isna()]

In [63]:
missing_df

age_group                        type       injury_mechanism  number_est  \
4        0-17  Emergency Department Visit  Intentional self-harm         NaN   
11        0-4  Emergency Department Visit  Intentional self-harm         NaN   
18       5-14  Emergency Department Visit  Intentional self-harm         NaN   
60      65-74  Emergency Department Visit  Intentional self-harm         NaN   
67        75+  Emergency Department Visit  Intentional self-harm         NaN   
81       0-17            Hospitalizations  Intentional self-harm         NaN   
88        0-4            Hospitalizations  Intentional self-harm         NaN   
95       5-14            Hospitalizations  Intentional self-harm         NaN   
158      0-17                      Deaths  Intentional self-harm         NaN   
165       0-4                      Deaths  Intentional self-harm         NaN   
172      5-14                      Deaths  Intentional self-harm         NaN   

     rate_est  
4         NaN  
11        NaN  
18        NaN  
60        NaN  
67        NaN  
81        NaN  
88        NaN  
95        NaN  
158       NaN  
165       NaN  
172       NaN

#### tbi_year.csv 
Read the raw data in from tidytuesday site.

In [64]:
data_url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-03-24/tbi_year.csv'

In [65]:
year_df = pd.read_csv(data_url)

#### Data description
This file specifies the number of TBIs and the rate per 100,000 by year, type of measure, injury mechanism from 2006 to 2014. There are 216 observations and five variables: injury_mechanism, type (type of measure), year, number_est, and rate_est. The Class for year in the data dictionary is incorrect - year is of type integer not character. The description for number_est in the data dictionary is incorrect - it should say 'Rate/100,000 in the specified year.' 

There are 27 missing values in the number_est column. These missing values occur where injury_mechanism equals 'Total' for each of the three types of measures: emergency dept, hospital, and deaths. I am thinking this value is the sum of all TBIs by any injury mechanism for the given year, and these missing values could be calculated from the data.

In [66]:
year_df.shape

(216, 5)

In [67]:
year_df.head(30)

injury_mechanism  \
0                               Motor vehicle crashes   
1                               Motor vehicle crashes   
2                               Motor vehicle crashes   
3                               Motor vehicle crashes   
4                               Motor vehicle crashes   
5                               Motor vehicle crashes   
6                               Motor vehicle crashes   
7                               Motor vehicle crashes   
8                               Motor vehicle crashes   
9                                 Unintentional falls   
10                                Unintentional falls   
11                                Unintentional falls   
12                                Unintentional falls   
13                                Unintentional falls   
14                                Unintentional falls   
15                                Unintentional falls   
16                                Unintentional falls   
17                                Unintentional falls   
18     Unintentionally struck by or against an object   
19     Unintentionally struck by or against an object   
20     Unintentionally struck by or against an object   
21     Unintentionally struck by or against an object   
22     Unintentionally struck by or against an object   
23     Unintentionally struck by or against an object   
24     Unintentionally struck by or against an object   
25     Unintentionally struck by or against an object   
26     Unintentionally struck by or against an object   
27  Other unintentional injury, mechanism unspecified   
28  Other unintentional injury, mechanism unspecified   
29  Other unintentional injury, mechanism unspecified   

                          type  year  rate_est  number_est  
0   Emergency Department Visit  2006      85.3    254793.0  
1   Emergency Department Visit  2007      83.8    252459.0  
2   Emergency Department Visit  2008      83.9    254391.0  
3   Emergency Department Visit  2009      88.7    270240.0  
4   Emergency Department Visit  2010      95.3    292942.0  
5   Emergency Department Visit  2011      98.7    305694.0  
6   Emergency Department Visit  2012      99.9    311271.0  
7   Emergency Department Visit  2013      99.6    312048.0  
8   Emergency Department Visit  2014     106.0    334144.0  
9   Emergency Department Visit  2006     208.8    625098.0  
10  Emergency Department Visit  2007     221.7    672504.0  
11  Emergency Department Visit  2008     240.9    740528.0  
12  Emergency Department Visit  2009     298.8    928049.0  
13  Emergency Department Visit  2010     316.3    980245.0  
14  Emergency Department Visit  2011     348.7   1090575.0  
15  Emergency Department Visit  2012     355.9   1120041.0  
16  Emergency Department Visit  2013     362.9   1163473.0  
17  Emergency Department Visit  2014     374.9   1213476.0  
18  Emergency Department Visit  2006      90.8    266268.0  
19  Emergency Department Visit  2007      90.6    267175.0  
20  Emergency Department Visit  2008      95.8    284297.0  
21  Emergency Department Visit  2009     123.9    369348.0  
22  Emergency Department Visit  2010     127.0    379601.0  
23  Emergency Department Visit  2011     138.8    415408.0  
24  Emergency Department Visit  2012     149.8    447591.0  
25  Emergency Department Visit  2013     141.0    423477.0  
26  Emergency Department Visit  2014     143.9    433951.0  
27  Emergency Department Visit  2006      40.4    118834.0  
28  Emergency Department Visit  2007      42.7    126635.0  
29  Emergency Department Visit  2008      43.1    129046.0

Display missing observations and data types for each variable.

In [68]:
year_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 216 entries, 0 to 215
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   injury_mechanism  216 non-null    object 
 1   type              216 non-null    object 
 2   year              216 non-null    int64  
 3   rate_est          216 non-null    float64
 4   number_est        189 non-null    float64
dtypes: float64(2), int64(1), object(2)
memory usage: 8.6+ KB


In [69]:
print(year_df.isna().sum())

injury_mechanism     0
type                 0
year                 0
rate_est             0
number_est          27
dtype: int64


Display number of unique observations per variable and display value counts for certain variables.

In [70]:
year_df.nunique()

injury_mechanism      8
type                  3
year                  9
rate_est            159
number_est          188
dtype: int64

In [71]:
year_df.type.value_counts()

Deaths                        72
Hospitalizations              72
Emergency Department Visit    72
Name: type, dtype: int64

In [75]:
year_df.year.value_counts()

2006    24
2007    24
2008    24
2009    24
2010    24
2011    24
2012    24
2013    24
2014    24
Name: year, dtype: int64

Look at missing data to determine if there is a pattern.

In [76]:
missing_df = year_df.loc[year_df['number_est'].isna()]

In [77]:
missing_df.head(27)

injury_mechanism                        type  year  rate_est  number_est
63             Total  Emergency Department Visit  2006     521.6         NaN
64             Total  Emergency Department Visit  2007     533.8         NaN
65             Total  Emergency Department Visit  2008     559.8         NaN
66             Total  Emergency Department Visit  2009     670.2         NaN
67             Total  Emergency Department Visit  2010     699.5         NaN
68             Total  Emergency Department Visit  2011     756.3         NaN
69             Total  Emergency Department Visit  2012     771.9         NaN
70             Total  Emergency Department Visit  2013     784.6         NaN
71             Total  Emergency Department Visit  2014     801.9         NaN
135            Total            Hospitalizations  2006      92.2         NaN
136            Total            Hospitalizations  2007      87.2         NaN
137            Total            Hospitalizations  2008      85.8         NaN
138            Total            Hospitalizations  2009      87.9         NaN
139            Total            Hospitalizations  2010     102.5         NaN
140            Total            Hospitalizations  2011      82.5         NaN
141            Total            Hospitalizations  2012      88.4         NaN
142            Total            Hospitalizations  2013      84.2         NaN
143            Total            Hospitalizations  2014      84.9         NaN
207            Total                      Deaths  2006      17.9         NaN
208            Total                      Deaths  2007      17.8         NaN
209            Total                      Deaths  2008      17.2         NaN
210            Total                      Deaths  2009      16.6         NaN
211            Total                      Deaths  2010      16.5         NaN
212            Total                      Deaths  2011      16.6         NaN
213            Total                      Deaths  2012      16.8         NaN
214            Total                      Deaths  2013      16.8         NaN
215            Total                      Deaths  2014      16.8         NaN

#### tbi_military.csv 
Read the raw data in from tidytuesday site.

In [78]:
data_url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-03-24/tbi_military.csv'

In [79]:
military_df = pd.read_csv(data_url)

## Data description
This file specifies the number of TBIs sustained by military personnel and the rate per 100,000 by service, component, severity, and year from 2006 to 2014. There are 450 observations and five variables: service (branch), component (duty type), severity, diagnosed, and year. 

There are 12 missing values in the diagnosed column. There is no pattern to the missing values.

In [80]:
military_df.shape

(450, 5)

In [81]:
military_df.head(20)

service component          severity  diagnosed  year
0     Army    Active       Penetrating      189.0  2006
1     Army    Active            Severe      102.0  2006
2     Army    Active          Moderate      709.0  2006
3     Army    Active              Mild     5896.0  2006
4     Army    Active  Not Classifiable      122.0  2006
5     Army     Guard       Penetrating       33.0  2006
6     Army     Guard            Severe       26.0  2006
7     Army     Guard          Moderate      177.0  2006
8     Army     Guard              Mild     1332.0  2006
9     Army     Guard  Not Classifiable       29.0  2006
10    Army   Reserve       Penetrating       12.0  2006
11    Army   Reserve            Severe       11.0  2006
12    Army   Reserve          Moderate       63.0  2006
13    Army   Reserve              Mild      541.0  2006
14    Army   Reserve  Not Classifiable       12.0  2006
15    Navy    Active       Penetrating       29.0  2006
16    Navy    Active            Severe       28.0  2006
17    Navy    Active          Moderate      217.0  2006
18    Navy    Active              Mild     2143.0  2006
19    Navy    Active  Not Classifiable       56.0  2006

Display missing observations and data types for each variable.

In [82]:
military_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 450 entries, 0 to 449
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   service    450 non-null    object 
 1   component  450 non-null    object 
 2   severity   450 non-null    object 
 3   diagnosed  438 non-null    float64
 4   year       450 non-null    int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 17.7+ KB


In [83]:
print(military_df.isna().sum())

service       0
component     0
severity      0
diagnosed    12
year          0
dtype: int64


Display number of unique observations per variable and display value counts for certain variables.

In [84]:
military_df.nunique()

service        4
component      3
severity       5
diagnosed    229
year           9
dtype: int64

In [85]:
military_df.year.value_counts()

2006    50
2007    50
2008    50
2009    50
2010    50
2011    50
2012    50
2013    50
2014    50
Name: year, dtype: int64

In [86]:
military_df.severity.value_counts()

Mild                90
Severe              90
Not Classifiable    90
Penetrating         90
Moderate            90
Name: severity, dtype: int64

Look at rows for one severity level and one year to verify military data is provided by service, component, severity, and year.

In [87]:
pen_df = military_df.loc[(military_df['severity'] == 'Penetrating') & (military_df['year'] == 2009)]

In [88]:
pen_df.shape

(10, 5)

In [89]:
pen_df.head(10)

service component     severity  diagnosed  year
150       Army    Active  Penetrating      333.0  2009
155       Army     Guard  Penetrating       67.0  2009
160       Army   Reserve  Penetrating       17.0  2009
165       Navy    Active  Penetrating       25.0  2009
170       Navy   Reserve  Penetrating        3.0  2009
175  Air Force    Active  Penetrating       25.0  2009
180  Air Force     Guard  Penetrating        2.0  2009
185  Air Force   Reserve  Penetrating        2.0  2009
190    Marines    Active  Penetrating       42.0  2009
195    Marines   Reserve  Penetrating        2.0  2009

In [90]:
missing_df = military_df.loc[military_df['diagnosed'].isna()]

In [91]:
missing_df.head(12)

service component     severity  diagnosed  year
20        Navy   Reserve  Penetrating        NaN  2006
35   Air Force   Reserve  Penetrating        NaN  2006
85   Air Force   Reserve  Penetrating        NaN  2007
86   Air Force   Reserve       Severe        NaN  2007
95     Marines   Reserve  Penetrating        NaN  2007
131  Air Force     Guard       Severe        NaN  2008
135  Air Force   Reserve  Penetrating        NaN  2008
281  Air Force     Guard       Severe        NaN  2011
321       Navy   Reserve       Severe        NaN  2012
336  Air Force   Reserve       Severe        NaN  2012
370       Navy   Reserve  Penetrating        NaN  2013
386  Air Force   Reserve       Severe        NaN  2013

#### dod_tbi_2006.csv 
Read the raw data in from tidytuesday site.

I could not figure out how to access this file without downloading it and opening it up with Excel and TextEdit. When I did this, I found a lot of non-alpha-numeric characters. There were some lines in the file that looked like tab separated data, but there were no headings and there were not many lines of this data compared to the number of lines of unreadable characters. I do not think this file was meant to be accessed for anything - perhaps it has a very specific purpose only known to the creators.

In [94]:
data_url = 'dod_tbi_2006.csv'

In [95]:
dod_2006_df = pd.read_csv(data_url)

FileNotFoundError: [Errno 2] No such file or directory: 'dod_tbi_2006.csv'